### PR TITLE
Write chunk metadata as header in chunks, not into index

### LINKED ISSUE #7: Chunk store index shouldn't need to include chunk metadata
_From @tomwilkie_

_Copied from original issue: tomwilkie/frankenstein#6_

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #7 and lays ground work for #130

### CODE REVIEW

**Comment by tomwilkie** [comment]
Tested locally and all seems to work.

**Comment by tomwilkie** [comment]
@jml split into two PRs: first is just vendoring, second if code.

File: chunk/chunk.go [hunk 1]
```diff
@@ -23,11 +40,82 @@ func (cs ByID) Len() int           { return len(cs) }
...
+
+	err := json.NewDecoder(snappy.NewReader(&io.LimitedReader{
+		N: int64(metadataLen),
+		R: r,
+	})).Decode(c)
+	if err != nil {
+		return err
+	}
+
+	if c.Encoding == prom_chunk.Delta {
```
Comment (jml) [inline chunk/chunk.go]: I think the comment you made in the struct (about never using Delta) should be moved down here.

File: chunk/chunk_store_test.go [hunk 2]
```diff
@@ -75,7 +79,8 @@ func TestChunkStore(t *testing.T) {
 			"bar":  "beep",
 			"toms": "code",
 		},
-		Data: []byte{},
+		Encoding: chunk.DoubleDelta,
+		Data:     chunks[0],
```
Comment (jml) [inline chunk/chunk_store_test.go]: I'd be a lot happier if there were direct tests that showed that `reader` and `decode` were the reverse of each other. Also if there were tests that exercised the `metadataInIndex` path.

**Comment by tomwilkie** [comment]
Just as well you asked @jml!  I hit the following bug in the docker library I vendored in:

https://github.com/docker/docker/pull/28552

File: chunk/chunk_store_test.go (hunk 2 above)
Comment (tomwilkie) [inline chunk/chunk_store_test.go]: Added test.  Not going to exercise metadataInIndex, as I'm going to remove that code at some point.

**Comment by tomwilkie** [comment]
Forgot to `git add` test.  Done now.  Also docker merged that PR so I updated the vendoring.

**Comment by jml** [comment]
Thanks!

File: chunk/chunk.go [hunk 3]
```diff
@@ -23,11 +37,97 @@ func (cs ByID) Len() int           { return len(cs) }
...
+
+	err := json.NewDecoder(snappy.NewReader(&io.LimitedReader{
+		N: int64(metadataLen),
+		R: r,
+	})).Decode(c)
+	if err != nil {
+		return err
+	}
+
+	// Flag indicates if metadata was written to index, and if false implies
```
Comment (juliusv) [inline chunk/chunk.go]: Is this comment in the wrong place? Should it belong to the `c.metadataInIndex` check?

Comment (tomwilkie) [inline chunk/chunk.go]: Yeah looks like it belongs up at the top; this function was factored out.

**Pull request merged**